slycot/transform.py

```diff
@@ -333,6 +333,254 @@ def tb04ad(n,m,p,A,B,C,D,tol1=0.0,tol2=0.0,ldwork=None):
+    >>> B = np.array([[0.],[100]])
+    >>> C = np.array([[1., 0.]])
+    >>> n = np.shape(A)[0]
+    >>> m = np.shape(B)[1]
+    >>> p = np.shape(C)[0]
+    >>> jw_s = [1j*10, 1j*15]
+    >>> at, bt, ct, g_1, hinvb, info = slycot.tb05ad(n, m, p, jw_s[0],
+                                                    A, B, C, job='NG')
+    >>> g_2, hinv2,info = slycot.tb05ad(n, m, p, jw_s[1], at, bt, ct, job='NH')
+
+    """
+    def error_handler(out, arg_list):
+        if out[-1] < 0:
+            # Conform fortran 1-based argument indexing to
+            # to python zero indexing.
+            error_text = ("The following argument had an illegal value: "
+                          + arg_list[-out[-1]-1])
+            e = ValueError(error_text)
+            e.info = out[-1]
+            raise e
+        if out[-1] == 1:
+            error_text = ("More than 30*"+str(n)+"iterations are required "
+                          "to iso late the eigenvalue of A; the computations "
+                          "are continued.")
+            e = ValueError(error_text)
+            e.info = out[-1]
+            raise e
+        if out[-1] == 2:
+            error_text = ("Either FREQ is too near to an eigenvalue of A, or "
+                          "RCOND is less than the machine precision EPS.")
+            e = ValueError(error_text)
+            e.info = out[-1]
+            raise e
+
+    hidden = ' (hidden by the wrapper)'
+    arg_list = ['baleig'+hidden, 'inita'+hidden, 'n', 'm', 'p', 'freq', 'a',
+                'lda'+hidden, 'b', 'ldb'+hidden, 'c', 'ldc'+hidden, 'rcond',
+                'g', 'ldg'+hidden, 'evre', 'evim', 'hinvb', 'ldhinv'+hidden,
+                'iwork'+hidden, 'dwork'+hidden, 'ldwork'+hidden,
+                'zwork'+hidden, 'lzwork'+hidden, 'info'+hidden]
+    # Fortran function prototype:
+    # TB05AD(baleig,inita,n,m,p,freq,a,lda,b,ldb,c,ldc,rcond,g,ldg,evre,evim,hinvb,ldhinv,
+    # iwork,dwork,ldwork,zwork,lzwork,info)
+
+    # Sanity check on matrix dimensions
+    if A.shape != (n, n):
+        e = ValueError("The shape of A is (" + str(A.shape[0]) + "," +
+                       str(A.shape[1]) + "), but expected (" + str(n) +
+                       "," + str(n) + ")")
+        e.info = -6  
+        raise e
+
+    if B.shape != (n, m):
+        e = ValueError("The shape of B is (" + str(B.shape[0]) + "," +
+                       str(B.shape[1]) + "), but expected (" + str(n) +
+                       "," + str(m) + ")")
+        e.info = -8
+        raise e
+    if C.shape != (p, n):
+        e = ValueError("The shape of C is (" + str(C.shape[0]) + "," +
+                       str(C.shape[1]) + "), but expected (" + str(p) +
+                       "," + str(n) + ")")
+        e.info = -10
+        raise e
+
+    # ----------------------------------------------------
+    # Checks done, do computation.
+    if job == 'AG':
+        out = _wrapper.tb05ad_ag(n, m, p, jomega, A, B, C)
+        error_handler(out, arg_list)
```

Comment:
This is good; it's tidier than the error handling elsewhere in Slycot.  

It might be good to turn your `error_handler` into the constructor for (currently non-existant) `SlycotValueError`, inherited from `ValueError`, that handle alls the re-indexing, etc., in one place.  However, that's for another PR.